dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/reservedvalue/hibernate/HibernateReservedValueStore.java

```diff
@@ -71,21 +71,58 @@ public HibernateReservedValueStore( SessionFactory sessionFactory, JdbcTemplate
     }
 
     @Override
-    public List<ReservedValue> reserveValuesAndCheckUniqueness( ReservedValue reservedValue,
+    public List<ReservedValue> getAvailableValues( ReservedValue reservedValue,
         List<String> values )
     {
         List<String> availableValues = getIfAvailable( reservedValue, values );
 
-        List<ReservedValue> toAdd = availableValues.stream()
+        return availableValues.stream()
             .map( value -> reservedValue.toBuilder().value( value ).build() ).collect( Collectors.toList() );
+    }
 
+    @Override
+    public void bulkInsertReservedValues( List<ReservedValue> toAdd )
+    {
         BatchHandler<ReservedValue> batchHandler = batchHandlerFactory
             .createBatchHandler( ReservedValueBatchHandler.class ).init();
 
         toAdd.forEach( batchHandler::addObject );
         batchHandler.flush();
+    }
 
-        return toAdd;
+    private List<String> getIfAvailable( ReservedValue reservedValue, List<String> values )
+    {
+        List<String> reservedValues = getReservedValues( reservedValue, values ).stream()
+            .map( ReservedValue::getValue )
+            .collect( Collectors.toList() );
+
+        values.removeAll( reservedValues );
+
+        Optional.of( values ).filter(
+            v -> !v.isEmpty() && Objects.valueOf( reservedValue.getOwnerObject() ).equals( TRACKEDENTITYATTRIBUTE ) )
+            .ifPresent(
+                v -> values.removeAll( getUntypedSqlQuery(
+                    "SELECT value FROM trackedentityattributevalue WHERE trackedentityattributeid = (SELECT trackedentityattributeid FROM trackedentityattribute WHERE uid = :uid) AND value IN :values" )
```

Comment:
The TEAV table (tracked entity attribute value) is potentially the slowest to look up values in on bigger instances. If we are looking them up, we have an index for lower(value), to speed it up.